[BugFix] [RHEL/7] [Fedora] Fix for issue #851

iankko opened the pull request:
Update the form of ```disable_interactive_boot``` rule for Systemd (RHEL/7, Fedora) based systems

Update:
* rule title (since on RHEL/7, Fedora systems interactive boot is disabled by default),
* XCCDF prose to mention updating ```systemd.confirm_spawn=(1|yes|true|on)``` kernel argument in ```/etc/default/grub``` (rather than setting ```PROMPT=no``` in ```/etc/sysconfig/init```),
* OVAL checks for RHEL/7 and Fedora (to check ```systemd.confirm_spawn``` in ```/etc/default/grub```),
* Finally also update remediation script for this rule for RHEL/7 and  Fedora

Additional information:
* See https://access.redhat.com/solutions/2112701 for further details

Fixes:
* https://github.com/OpenSCAP/scap-security-guide/issues/851

Please review.

Thank you, Jan

iankko commented:
Also fixes downstream: https://bugzilla.redhat.com/show_bug.cgi?id=1291516

redhatrises commented on Fedora/input/xccdf/system/accounts/physical.xml:
```diff
@@ -207,27 +207,38 @@ loss of availability of systems due to unintentional reboot.
-interactive system startup, in which it is possible to select the
-set of services which are started on boot.
+Fedora systems support an "interactive boot" option that can be used to prevent
+services from being started. On Fedora system interactive boot can be enabled
```
nitpick: missing `a` and `,` so it should be `On a Fedora system,`

redhatrises commented on RHEL/7/input/xccdf/system/accounts/physical.xml:
```diff
@@ -224,27 +224,38 @@ the non-graphical <tt>multi-user.target</tt> mode.
-set of services which are started on boot.
+Red Hat Enterprise Linux systems support an "interactive boot" option that can
+be used to prevent services from being started. On Red Hat Enterprise Linux 7
+system interactive boot can be enabled by providing some of <tt>1</tt>,
```
nitpick: missing `a` and `,` so it should be `On a Red Hat Enterprise Linux 7 system,`

redhatrises replied:
@iankko nitpick: change `some of` to `a`

iankko replied:
Thanks for these nitpicks (since not sure what's the proper form in English)!. Will fix them.

redhatrises commented on Fedora/input/xccdf/system/accounts/physical.xml:
```diff
@@ -207,27 +207,38 @@ loss of availability of systems due to unintentional reboot.
-set of services which are started on boot.
+Fedora systems support an "interactive boot" option that can be used to prevent
+services from being started. On Fedora system interactive boot can be enabled
+by providing some of <tt>1</tt>, <tt>yes</tt>, <tt>true</tt>, or <tt>on</tt>
```
@iankko nitpick: change `some of` to `a`

redhatrises commented on shared/oval/disable_interactive_boot.xml:
```diff
@@ -1,34 +1,74 @@
+        <criterion test_ref="test_disable_interactive_boot_grub_cmdline_linux_default"
+        comment="Check systemd.confirm_spawn=(1|yes|true|on) not in GRUB_CMDLINE_LINUX_DEFAULT" />
+        <criterion test_ref="test_bootloader_recovery_disabled"
+        comment="Check GRUB_DISABLE_RECOVERY=true in /etc/default/grub" />
```
Also, not sure that we should check for `GRUB_DISABLE_RECOVERY` in every Grub OVAL that is generated, or should we?

iankko replied:
The issue is ```GRUB_CMDLINE_LINUX``` value is used only when ```GRUB_DISABLE_RECOVERY``` is not enabled. If it is, ```GRUB_CMDLINE_LINUX_DEFAULT``` value is honored instead. This has been discussed earlier within som PR related to ```nousb``` grub.cfg kernel argument IIRC.

redhatrises commented on shared/oval/disable_interactive_boot.xml:
```diff
@@ -1,34 +1,74 @@
-           use case on RHEL-7 and Fedora per:
-             https://rhsummit.files.wordpress.com/2014/04/summit_demystifying_systemd1.pdf Page #20
-           SSG Upstream ticket to come yet -->
+    <criteria operator="OR">
```
@iankko shouldn't this be `"AND"` instead of `"OR"` because if you do not have a `GRUB_CMDLINE_LINUX_DEFAULT` line in `/etc/default/grub` but have `GRUB_CMDLINE_LINUX` with systemd.confirm_spawn=true this will evaluate as true rather than false, no?

iankko replied:
@redhatrises 
This seems to be best explanation of the slight difference I was able to find: http://askubuntu.com/questions/575651/what-is-the-difference-between-grub-cmdline-linux-and-grub-cmdline-linux-default/780545#780545

As I understand it, from https://www.gnu.org/software/grub/manual/grub.html being:
* if GRUB_DISABLE_RECOVERY=true, the default kernel menu entry will be equipped also with options / settings from GRUB_CMDLINE_LINUX_DEFAULT (they would be appended after values / settings from GRUB_CMDLINE_LINUX)
* if GRUB_DISABLE_RECOVERY=false, just GRUB_CMDLINE_LINUX settings would be effective.

So IMHO we should:
* if GRUB_DISABLE_RECOVERY=false inspect just GRUB_CMDLINE_LINUX settings,
* if GRUB_DISABLE_RECOVERY=true inspect both GRUB_CMDLINE_LINUX and GRUB_CMDLINE_LINUX_DEFAULT settings (will recheck the OVAL if it's working as described here actually).

iankko commented:
@redhatrises 
Fixed the text typos reported earlier. If you want me to squash the second commit into the first one, let me know (will do that).

For the ```GRUB_DISABLE_RECOVERY``` question -- have reviewed the OVAL and looks reasonable to me (IOW case of ```GRUB_CMDLINE_LINUX``` would be inspected always. In the case also ```GRUB_DISABLE_RECOVERY==true```, also GRUB_CMDLINE_LINUX_DEFAULT would be checked). Am I missing something?

redhatrises commented:
@iankko in testing when I have `systemd.confirm_spawn=true` in the `GRUB_CMDLINE_LINUX` line, the check passes with true instead of false.

iankko commented:
@redhatrises 
Will have a look yet.

iankko commented:
@redhatrises 
You are correct. The top-level criterion operator should be "AND" instead of "OR". Did test with GRUB_DISABLE_RECOVERY=false before. But after enabling it, I can reproduce the issue you mention above.

Changed the top-level operator from "OR" to "AND" fixed it (now it reports fail if systemd.confirm_spawn is present in GRUB_CMDLINE_LINUX). Added another commit and squashed the previous changes into one.

In the light of the above, looks we should revisit other grub2 grub.cfg GRUB_CMDLINE_LINUX{,_DEFAULT} tests too (IIRC there are 2-3 more of them yet).

iankko commented:
Should be safe to rereview once Jenkins tests finished.

redhatrises commented:
Thanks @iankko!

> In the light of the above, looks we should revisit other grub2 grub.cfg GRUB_CMDLINE_LINUX{,_DEFAULT} tests too (IIRC there are 2-3 more of them yet).

Let's open a ticket for this to keep track for this unless you have the time to get a PR fix for it today.

redhatrises commented:
Ack.

iankko commented:
@redhatrises 
> Let's open a ticket for this to keep track for this unless you have the time to get a PR fix for it today.

Has run out of time today (looking in parallel at something else yet), but will file them tomorrow (I still have got to file those three tickets related with ```shared/xccdf/remediation_functions.xml``` enhancements), thus will file them all in one batch.

Thanks for the review!

iankko commented:
@redhatrises 
> > Thanks @iankko!
> >  In the light of the above, looks we should revisit other grub2 grub.cfg
> > GRUB_CMDLINE_LINUX{,_DEFAULT} tests too (IIRC there are 2-3 more of them yet).
> >
> Let's open a ticket for this to keep track for this unless you have the time to get a PR fix for it today.

The need to inspect remaining ```/etc/default/grub``` ```RHEL/7``` and ```Fedora``` OVALs wrt to GRUB_CMDLINE_LINUX / GRUB_CMDLINE_LINUX_DEFAULT is now tracked under:
* https://github.com/OpenSCAP/scap-security-guide/issues/1307

Sorry for the delay.
